shopinvader_restapi/services/cart.py

```diff
@@ -0,0 +1,533 @@
+        """
+        This service allows
+        :param params:
+        :return:
+        """
+        return self._copy(**params)
+
+    def _copy(self, **params):
+        """
+        Copy the cart given by id without the lines
+        They will be re-added
+        :return: dict/json
+        """
+        cart = self.env["sale.order"].search(
+            self._get_validator_cart_by_id_domain(params.get("id"))
+        )
+        # Copy the existing cart
+        # Delete all lines and re-add them with 'shopinvader' flavour
+        new_cart = cart.copy({"order_line": False, "typology": "cart"})
+        for line in self._get_lines_to_copy(cart):
+            vals = self._get_line_copy_vals(line)
+            self._add_item(new_cart, vals)
+
+        return self._to_json(new_cart)
+
+    # Validator
+    def _cart_validator_exists(self, field, value, error):
+        """
+        Implements 'check_with' validation
+        :param field:
+        :param value:
+        :param error:
+        :return:
+        """
+        cart = self.env["sale.order"].search(
+            self._get_validator_cart_by_id_domain(value)
+        )
+        if len(cart) != 1:
+            error(field, _("The cart does not exists or does not belong to you!"))
+
+    def _validator_copy(self):
+        return {
+            "id": {
+                "coerce": to_int,
+                "required": True,
+                "type": "integer",
+                "check_with": self._cart_validator_exists,
+            }
+        }
+
+    def _validator_return_copy(self):
+        return Validator({}, allow_unknown=True)
+
+    def _validator_search(self):
+        return {}
+
+    def _validator_clear(self):
+        return {}
+
+    def _subvalidator_shipping(self):
+        return {
+            "type": "dict",
+            "schema": {
+                "address": {
+                    "type": "dict",
+                    "schema": {"id": {"coerce": to_int}},
+                }
+            },
+        }
+
+    def _subvalidator_invoicing(self):
+        return {
+            "type": "dict",
+            "schema": {
+                "address": {
+                    "type": "dict",
+                    "schema": {"id": {"coerce": to_int}},
+                }
+            },
+        }
+
+    def _subvalidator_step(self):
+        return {
+            "type": "dict",
+            "schema": {
+                "current": {"type": "string"},
+                "next": {"type": "string"},
+            },
+        }
+
+    def _validator_update(self):
+        return {
+            "client_order_ref": {
+                "type": "string",
+                "required": False,
+                "nullable": True,
+            },
+            "step": self._subvalidator_step(),
+            "shipping": self._subvalidator_shipping(),
+            "invoicing": self._subvalidator_invoicing(),
+            "note": {"type": "string"},
+        }
+
+    def _validator_add_item(self):
```

Comment:
Return validators are missing for `add_item`, `update_item` and `delete_item`. They are now almost required in v16, as suggested in logs:

 WARNING odoo-test odoo.addons.base_rest.components.service: DEPRECATED: You must define an output schema for method add_item in service shopinvader.cart.service